File: client/src/main/java/au/com/codeka/warworlds/client/concurrency/GmsTask.java
package au.com.codeka.warworlds.client.concurrency;

/** Simple wrapper around GmsCore tasks API. */
public class GmsTask<R> extends Task<Void, R> {
  public GmsTask(TaskRunner taskRunner, com.google.android.gms.tasks.Task<R> gmsTask) {
    super(taskRunner);
    gmsTask.addOnFailureListener(this::onError);
    gmsTask.addOnCompleteListener(task -> {
      try {
        this.onComplete(task.getResult());
      } catch (Exception e) {
        onError(e);
      }
    });
  }
}

File: client/src/main/java/au/com/codeka/warworlds/concurrency/GmsTask.kt
package au.com.codeka.warworlds.concurrency

/** Simple wrapper around GmsCore tasks API.  */
class GmsTask<R>(taskRunner: TaskRunner, gmsTask: com.google.android.gms.tasks.Task<R>):
    Task<Void, R>(taskRunner) {
  init {
    gmsTask.addOnFailureListener { error: Exception? -> onError(error) }
    gmsTask.addOnCompleteListener { task: com.google.android.gms.tasks.Task<R> ->
      try {
        onComplete(task.result)
      } catch (e: Exception) {
        onError(e)
      }
    }
  }
}
